### PR TITLE
All sampled images in image list

### PR DESCRIPTION
**Pull request by PaulHax**
Now, the "image list" shows all images sampled from the dataset.  As you scroll down the table, images are loaded, transformed, and object detected.  To know what images are in view within the table, the `quasar.QTable` component has some "virtual scrolling" features that trigger a callback with the  "in view" row numbers.

When you select images from the embeddings plot, the app filters the image list.  Before, you had to select images from the embeddings plot to show them in the image list.

`images.py` caches loaded images, transformed images, and their object detection results.  `images.py` uses a least recently used cache strategy.  

To seperate caching method from downstream processing,  `embeddings_extractor.py` and `object_detector.py` now take only loaded `PIL.Image`s as arguments.

### CODE REVIEW

**Comment by PaulHax** [comment]
I looked at using `functools.lru_cache` for the `images.get_image` function.  `image.py` code runs some side-effects when a image is dropped from the cache (deletes image frome trame state).  I did not see a "item ejected" callback for `lru_cache`.

**Comment by vicentebolea** [comment]
@PaulHax it seems that the current BufferCache does not remove the item from the trame cache, in any case I believe that we should detach the cache from the trame state, the evicting instruction is `self.cache.popitem(last=False)`. lru_cache decorator should be able to substitute this.

**Comment by vicentebolea** [comment]
@PaulHax looks good, I want to better understand this, we went through the changes a couple of weeks ago but I Want to review this in more detail. I wonder if you could edit the description to add a brief note of the feature, the reason for adding the proposed new modules and why we want to change the existing "library" modules.

**Comment by PaulHax** [comment]
> it seems that the current BufferCache does not remove the item from the trame cache, 

Holy mem leak Batman!  Good catch.  I fix in PR cuz i can't push to this branch anymore =)

If you have an idea on how to use `lru_cache` as user scrolls through long image list i'm all ears.  I still don't know how that would work.

**Comment by vicentebolea** [comment]
> If you have an idea on how to use lru_cache as user scrolls through long image list i'm all ears. I still don't know how that would work.

How about wrapping the image into a class that implements the __del__ method, the class also has a ref to the state. In this manner the imagewrapper class can deallocate itself from the state just before being destroyed.

**Comment by PaulHax** [comment]
Refactored `images.py` to use `@lru_cache` for loading and tranforming images with a wrapper the `__del__` trick. 🎉

**Review by vicentebolea — CHANGES_REQUESTED** [review]
The changes looks good to me, however, I am skeptical of moving data structures to singleton. I wrote a more detailed analysis in the comments among other comments.

I can see how this feature could be implemented without moving the `ImageManager` to a singleton implemented in image.py and I do not see any drawback other than having to instantiate in each application class when is it is not in a `standalone` mode.

I am also skeptical of having another annotations cache singleton too.

Other than that great work in this feature.

File: src/nrtk_explorer/app/images/images.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        """Remove a specific item from the cache."""
+        if key in self.cache:
+            self.on_clear_item(key)
+            del self.cache[key]
+
+    def clear(self):
+        """Clear the cache."""
+        for key in self.cache.keys():
+            self.on_clear_item(key)
+        self.cache.clear()
```
[the remark this answers is not in the document]
Comment (PaulHax) [reply src/nrtk_explorer/app/images/images.py:63]: How to use `@lru_cache` with the object detection function: 
```
def get_annotations(detector: ObjectDetector, id_to_image: Dict[str, Image.Image]):
```

File: src/nrtk_explorer/app/images/images.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            annotation_cache.clear_item(keys["transformed_image"])
+            delete_state(state, keys["meta_id"])
+
+    # create reactive annotation variables so ImageDetection component has live Refs
+    for id in new:
+        keys = get_image_state_keys(id)
+        for key in keys.values():
+            if not state.has(key):
+                state[key] = None
+    state.hovered_id = None
```
[the remark this answers is not in the document]
Comment (PaulHax) [reply src/nrtk_explorer/app/images/images.py]: In this commit: https://github.com/Kitware/nrtk-explorer/pull/88/commits/d3bd004737024e2e99821e8f4ab1f8a602c7f96b
I've factored out the annotation/obejct-detection code from `images.py`.  `annotations.py` computes/retreaves the annotations and manages the caches.  `stateful_annotations.py` copies the annotations to the trame state.

**Review by jourdain — APPROVED** [review]

**Pull request merged**